zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/TypedEventWriter.java

```diff
@@ -137,4 +140,25 @@ default void appendFollowUpEvent(
    * @return true if an event of length {@code eventLength} can be written
    */
   boolean canWriteEventOfLength(final int eventLength);
+
+  /**
+   * Validate that the provided key is valid: - it's assigned to the expected partition - it's not
+   * higher than the current key
+   *
+   * @param key to validate
+   * @param expectedPartitionId expected partition
+   * @param keyGenerator to verify the key is not exceeding the highest key generated
+   */
+  static void validateKey(
+      final long key, final int expectedPartitionId, final KeyGenerator keyGenerator) {
+    if (key == -1L) {
+      return;
+    }
+    final var decodedPartitionId = Protocol.decodePartitionId(key);
+    if (decodedPartitionId == expectedPartitionId && key > keyGenerator.getCurrentKey()) {
```

Comment:
❓ So it is valid if we write a follow-up event with a key from a different partition?

Reply:
Yeah I was wondering about this case: are all events expected to be on the same partition? Is it just commands that can have a key from another partition?

Reply:
Yeah I think this is valid, on deployment distribution, the receiving partition will write an event with the original key generated by the deployment partition.